src/gc.c

```diff
@@ -1771,11 +1771,41 @@ STATIC_INLINE uintptr_t gc_read_stack(void *_addr, uintptr_t offset,
     return *(uintptr_t*)real_addr;
 }
 
-JL_NORETURN NOINLINE void gc_assert_datatype_fail(jl_ptls_t ptls, jl_datatype_t *vt,
+JL_NORETURN NOINLINE void assert_validity_of_new_obj(jl_value_t* parent, jl_value_t* obj) {
+    jl_taggedvalue_t *o = jl_astaggedvalue(obj);
+    uintptr_t vtag = o->header & ~(uintptr_t)0xf;
+    jl_datatype_t *vt = (jl_datatype_t *)vtag;
+    //jl_datatype_t* vt = (jl_datatype_t*) jl_typeof(obj);
+    jl_datatype_t* parent_typ = (jl_datatype_t*) jl_typeof(parent);
+
+
+    // jl_safe_printf("Enqueuing object:\n");
+    // jl_(vt);
+
+    //if (__unlikely(!jl_is_datatype(vt) || vt->smalltag)) {
+    if (__unlikely(!jl_is_datatype(vt))) {
```

Comment:
CC: @d-netto and @vchuravy: I'm not sure why, but this is always triggering for me...

Reply:
Is this on top of master? If so #49556 bt @vtjnash has made this check more complicated.

Reply:
Yeah, it was on top of master.
Thanks.

I saw the assertion check was changed to `(__unlikely(!jl_is_datatype(vt) || vt->smalltag))` but removing the rhs of the `||` should make it trigger _less often_, not more often -- that's why i was confused.

Is there more that has to change for this?

Reply:
So is the idea something like this?:

This check here:
https://github.com/JuliaLang/julia/blob/31b1374ff632a5b767d1d6d873eb02d5b5288c0f/src/gc.c#L2569-L2573
was only happening as an _else block,_ after all these other possible cases. So it's not required that everything we push is a jl_value_t with a jl_datatype_t in its type tag - there are a bunch of other cases to check too.

Something like that?

Reply:
K, yeah, i think that's what's going on. Thanks for the pointer!
K, i'm just gonna checkout a commit from 6 days ago for now, then 😁

Reply:
Aha! It turned out the issue was the erroneous JL_NORETURN which i had accidentally copy/pasted from the function below. I've fixed that, and I also made the changes to account for Jameson's changes. Thank you!